asdf/extension/_compressor.py

```diff
@@ -0,0 +1,84 @@
+"""
+Compressor is an interface to implement extensions to the compression
+module. Extensions will typically subclass the Compressor ABC and
+provide that subclass as a setuptools entry point.  Decompressor
+is likewise used for decompression.
+
+Note that this interface has similar patterns to Converter.  This
+interface is designed for compression and decompression of ASDF
+binary array blocks, while Converter is designed for serialization
+of custom Python types into the YAML tree.
+"""
+
+import abc
+
+class Compressor(abc.ABC):
```

Comment:
The `asdf.compression` module also splits these into encoder vs decoder, but I wonder if that's necessary?  Maybe there are rare cases where someone would want to implement just one of decompress or compress, but that could be handled by implementing e.g. a compress method that raises `NotImplementedError`.

Reply:
Yeah, I initially started the implementation with a unified `Compressor` with `compress` and `decompress` methods.  I ran into problems in two places: the `Compressor` constructor, and methods that could be shared between the two, like `flush`.  In the constructor, you have to jump through hoops to split arguments into those passed to the underlying compression object and those passed to the decompression object (like `lzma.LZMA[De]Compressor`).  It's possible, but not particularly elegant.

I started doing that, but then I encountered the problem where `flush` can be present in both encoders and decoders, so you'd probably have to split that method into `flush_compress` and `flush_decompress`.  Again, it's possible, but I started to smell a pattern and so decided to try implementing them separately like the built-in compression had done.  I think it worked well for the `compression` module, and it also lets an `Extension` know what labels it can compress versus decompress, in the rare cases where those diverge.

But the extra surface area of `decompressors` in the `Extension` and elsewhere in the code is one of my least favorite parts of this PR.  I'm happy to work through the above issues if we think a unified `Compressor` would be better.

Reply:
I think there's something that I don't understand yet, why does there need to be a `flush` method?  Wouldn't `compress` and `decompress` carry through the operation from start to finish, including any necessary flushes?

I was imagining this library would pass the entirety of each array to `compress` and receive back a new array containing the entirety of the compressed data.  Maybe that doesn't work for memory management reasons?

Reply:
Yes, that's a good point.  On the compression side, there's probably not much use for `flush` since we always have the whole array data.  Even if we didn't (e.g. data was streaming in), we could do some buffering before compression.  The built-in compressors rely on `flush` being called, but since we're going to wrap those anyway, we can force a `flush` after each `compress` call, which shouldn't be too bad because our block size is big.

On the decompression side, the algorithms using block decompression (like lz4) should be fine without `flush`, as they will self-flush at the end of a block.  However, I think this will force us to read the entire array data before calling decompression like `zlib` that is not wrapped in a blocking scheme.  We can't call [`flush`](https://docs.python.org/3/library/zlib.html#zlib.Decompress.flush) after every file block because that renders the decompression object unusable.  So, we could permit a final `flush` call, but just for decompression instead of compression.  Or, we could force non-blocked decompression schemes to read the full array data, at the cost of some extra memory.

Reply:
I've been thinking about decompression a little more, and I think it may be reasonable to read the entirety of the compressed data before passing it to the plugin's `decompress` method.  This library memory-maps arrays by default, so under normal conditions we could do this in `asdf.compression.decompress`:

```python
compressed_data = fd.read_into_array(used_size)
```

without requiring any additional program memory.

Reply:
I think I am always suspicious of memory-mapped IO in performance-sensitive contexts because it defers the IO into the bowels of the computation, so the computation is constantly being interrupted waiting for IO.  And the IO is only being triggered sporadically when the computation needs the next section of data—except for read-ahead, which very well may eliminate this concern.  I can do some tests.

Reply:
I did a standalone experiment where I wrote 1 GB of Blosc-compressed random-float data to a network file system from one node, then went to another node and timed the IO + decompression of the data using memmap.  Then I wrote another 1 GB in a new file from the first node, and went back to the second node and timed IO + decompression using explicit IO, no memmap.  The latter was 2.5x faster than the former.

In detail, here's what I was timing:
with memmap:
```python
%timeit -n1 -r1 mcomp = np.memmap('/mnt/home/lgarrison/ceph/test1.dat'); decomp = blosc.decompress(mcomp)
# 6.47 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
```
no memmap:
```python
%timeit -n1 -r1 comp = open('/mnt/home/lgarrison/ceph/test2.dat', 'rb').read(); decomp = blosc.decompress(comp)
# 2.28 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
```

I repeated the experiment a dozen times, with consistent results.  Interestingly, when I increased the number of Blosc threads to 2, the memmap time decreased to 4 seconds; at 4 threads, 2.5 seconds; at 8 threads, 1.6 seconds (faster than explicit IO).  So I assume memmap is effectively parallelizing the IO, which results in a speedup, at least on a fast network file system.  The actual decompression (no IO) takes 0.8 seconds at 1 threads and 0.1 seconds at 8 threads, so we're IO dominated in all these tests.

So... I think these results are consistent with memmap's IO pattern being suboptimal for compression.  But on the flip side, using many threads can be even faster than explicit IO.  The explicit IO could be parallelized too, although memmap gives it for "free".  Overall, these results suggest to me that we should not rely on memmapped IO here, although I don't want to extrapolate to all systems based on my platform.

Here's the writing code if you want to repeat the test yourself.
<details>

```python
import asdf
import numpy as np
import blosc

def w(fn):
    rng = np.random.default_rng()
    a = rng.random(1*(1<<30)//4, dtype=np.float32)
    comp = blosc.compress(a.tobytes(), typesize=4, cname='zstd', clevel=1, shuffle=blosc.BITSHUFFLE)
    with open(f'/mnt/home/lgarrison/ceph/{fn}', 'wb') as fp:
        fp.write(comp)
```
</details>

The reason I used one node to write and one node to read was to avoid the file still being in the page cache from the writing.  I'm not on machines where I have root, so I can't manually drop the page cache.  It's possible that memmap performance is different somehow on a network system than on a local disk, but I also think a cold read from a network file system is a common ASDF use case.

Reply:
That's interesting. Thanks for doing this. Let us mull this over for a day or so. But I think it shows memmap has its issues.

Reply:
Wow, okay, yeah, that seems conclusive.  The two other options I can think of for avoiding multiple calls to decompress are 1) pass in a file-like object with a `read` method or 2) pass in a generator.  The file interface has the advantage of allowing the compressor to select its own block size, but we'd need to either wrap the real file so that it stops returning bytes once the (ASDF) block is consumed, or pass the limit as an additional argument.  Do either of those options strike you as a good idea?

Also let me know if you don't think a single call to decompress is a virtue worth pursuing.  I like the simplicity of a stateless compressor but I realize that's an aesthetic statement about an interface that end users will likely never see.